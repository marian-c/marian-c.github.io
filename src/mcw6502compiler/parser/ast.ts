// TODO: make these ints (remove the values)
export enum NodeKind {
  program = 'p',
  line = 'l',
  reference = 'r',
  instruction = 'i',
  directive = 'd',
  comment = 'c',
  valueLiteral = 'vl',
  labelDefinition = 'ld',
  symbolDefinition = 'sd',
  instructionOperator = 'or',
  instructionOperand = 'od',
  immAddressingMode = 'imm',
  absAddressingMode = 'abs',
  zp0AddressingMode = 'zp0',
  abxAddressingMode = 'abx',
  abyAddressingMode = 'aby',
  zpxAddressingMode = 'zpx',
  zpyAddressingMode = 'zpy',
  indAddressingMode = 'ind',
  izxAddressingMode = 'izx',
  izyAddressingMode = 'izy',
  relAddressingMode = 'rel',
}

export enum InstructionKind {
  withoutOperand,
  withOperand,
}

// DirectiveKind is also used as value, so don't change individual values
export enum DirectiveKind {
  org = 'org',
  word = 'word',
}

export enum ValueLiteralKind {
  ascii = 'ascii',
  bin = 'bin',
  dec = 'dec',
  hex = 'hex',
}

export interface NodeBase {
  readonly kind: NodeKind;
  readonly range: [number, number];
  readonly source?: string;
}
export interface InstructionOperatorNode extends NodeBase {
  readonly kind: NodeKind.instructionOperator;
  readonly value: string;
  readonly forcedTo?: 'b';
}

export interface ValueLiteralNode extends NodeBase {
  readonly kind: NodeKind.valueLiteral;
  readonly valueLiteralKind: ValueLiteralKind;
  readonly originalStringValue: string;
}
export interface ReferenceNode extends NodeBase {
  readonly kind: NodeKind.reference;
  readonly referencedName: string;
}

export type ValuableNode = ValueLiteralNode | ReferenceNode;

export interface ImmAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.immAddressingMode;
  readonly value: ValuableNode;
}

export interface AbsAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.absAddressingMode;
  readonly value: ValuableNode;
}

export interface Zp0AddressingModeNode extends NodeBase {
  readonly kind: NodeKind.zp0AddressingMode;
  readonly value: ValuableNode;
}

export interface AbxAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.abxAddressingMode;
  readonly value: ValuableNode;
}
export interface AbyAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.abyAddressingMode;
  readonly value: ValuableNode;
}
export interface ZpxAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.zpxAddressingMode;
  readonly value: ValuableNode;
}
export interface ZpyAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.zpyAddressingMode;
  readonly value: ValuableNode;
}
export interface IndAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.indAddressingMode;
  readonly value: ValuableNode;
}
export interface IzxAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.izxAddressingMode;
  readonly value: ValuableNode;
}
export interface IzyAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.izyAddressingMode;
  readonly value: ValuableNode;
}

export interface RelAddressingModeNode extends NodeBase {
  readonly kind: NodeKind.relAddressingMode;
  readonly value: ValuableNode;
}

export type AddressingModeNode =
  | ImmAddressingModeNode
  | AbsAddressingModeNode
  | Zp0AddressingModeNode
  | AbxAddressingModeNode
  | AbyAddressingModeNode
  | ZpxAddressingModeNode
  | ZpyAddressingModeNode
  | IndAddressingModeNode
  | IzxAddressingModeNode
  | IzyAddressingModeNode
  | RelAddressingModeNode;

export type InstructionOperandNode = AddressingModeNode;

export interface CommentNode extends NodeBase {
  readonly kind: NodeKind.comment;
  readonly value: string;
}

export interface InstructionWithoutOperandNode extends NodeBase {
  readonly kind: NodeKind.instruction;
  readonly instructionKind: InstructionKind.withoutOperand;
  readonly operator: InstructionOperatorNode;
}
export interface InstructionWithOperandNode extends NodeBase {
  readonly kind: NodeKind.instruction;
  readonly instructionKind: InstructionKind.withOperand;
  readonly operator: InstructionOperatorNode;
  readonly operand: InstructionOperandNode;
}

export interface LabelDefinitionNode extends NodeBase {
  readonly kind: NodeKind.labelDefinition;
  readonly value: string;
}

export type SymbolDefinitionValueNode = ValueLiteralNode;

export interface SymbolDefinitionNode extends NodeBase {
  readonly kind: NodeKind.symbolDefinition;
  readonly symbolName: string;
  readonly symbolValue: SymbolDefinitionValueNode;
}

export interface DirectiveNodeBase extends NodeBase {
  readonly kind: NodeKind.directive;
  readonly directiveKind: DirectiveKind;
}

export interface DirectiveOrgNode extends DirectiveNodeBase {
  readonly directiveKind: DirectiveKind.org;
  // TODO: future: could allow .org directives to reference
  //   symbols if they are resolved by now
  readonly address: ValueLiteralNode;
  // TODO: (fill) actually implement this
  // readonly fill?: ValuableNode;
}
export interface DirectiveWordNode extends DirectiveNodeBase {
  readonly directiveKind: DirectiveKind.word;
  readonly value: ValuableNode;
}

export type DirectiveNode = DirectiveOrgNode | DirectiveWordNode;

export type InstructionNode = InstructionWithoutOperandNode | InstructionWithOperandNode;

export type LineContentsNode =
  | InstructionNode
  | LabelDefinitionNode
  | SymbolDefinitionNode
  | DirectiveNode;

export interface LineNode extends NodeBase {
  kind: NodeKind.line;
  comment?: CommentNode | undefined;
  contents?: LineContentsNode;
}

export interface ProgramNode extends NodeBase {
  readonly kind: NodeKind.program;
  readonly lines: readonly LineNode[];
}
