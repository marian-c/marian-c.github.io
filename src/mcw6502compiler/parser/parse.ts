import { type Config, makeConfig } from '@/mcw6502compiler/config';
import {
  type ErrorResult,
  makeSuccess,
  type ParseResult,
  type Result,
  ResultKind,
} from '@/mcw6502compiler/parser/parserTypes';
import { type LineContentsNode, type LineNode, NodeKind } from '@/mcw6502compiler/parser/ast';
import { parseComment } from '@/mcw6502compiler/parser/parseComment';
import {
  labelDefinitionRegExp,
  parseLabelDefinition,
} from '@/mcw6502compiler/parser/parseLabelDefinition';
import { parseInstruction } from '@/mcw6502compiler/parser/parseInstruction';
import { emptyRegExp, symbolDefinitionLightRegExp } from '@/mcw6502compiler/parser/regExp';
import { parseDirective } from '@/mcw6502compiler/parser/parseDirective';
import { parseSymbolDefinition } from '@/mcw6502compiler/parser/parseSymbolDefinition';
import { expected, testRegExp } from '@/mcw6502compiler/parser/stringInput';

/**
 * Given a string (a source file) and some config, it will return the ParseResult
 */
export function parseSource(originalSource: string, initialConfig: Config): ParseResult {
  const config = makeConfig(initialConfig);

  const sourceNormalized = originalSource
    .replace('\r\n', '\n')
    .replace('\r', '\n')
    .replace('\t', ' ');
  const lines = sourceNormalized.split('\n');
  const meta = { originalLines: lines };
  // TODO: does it work with different line endings?
  let index = 0;
  const lineNodes: LineNode[] = [];
  let errors: ErrorResult[] = [];
  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed) {
      // is it a comment only line
      if (trimmed.startsWith(';')) {
        const commentResult = parseComment({ index, value: line, meta }, config);
        if (commentResult.kind === ResultKind.Error) {
          errors.push(commentResult);
        } else {
          lineNodes.push({
            kind: NodeKind.line,
            range: commentResult.value.range,
            comment: commentResult.value,
          });
        }
      } else {
        let result: Result<LineContentsNode>;
        // is it a label definition
        if (labelDefinitionRegExp.test(trimmed)) {
          result = parseLabelDefinition({ index, value: line, meta }, config);
        } else if (trimmed.startsWith('.')) {
          result = parseDirective({ index, value: line, meta }, config);
        } else if (symbolDefinitionLightRegExp.test(trimmed)) {
          result = parseSymbolDefinition({ index, value: line, meta }, config);
        } else {
          result = parseInstruction({ index, value: line, meta }, config);
        }
        const lineResultWithComment = (function addComment(
          r: Result<LineContentsNode>,
        ): Result<LineNode> {
          if (r.kind === ResultKind.Error) {
            return r;
          }
          let rest = r.rest;

          let commentResult;
          if (rest.value.trim().startsWith(';')) {
            let comment = parseComment(rest, config);
            if (comment.kind === ResultKind.Error) {
              return comment;
            }
            commentResult = comment;
            rest = commentResult.rest;
          }

          // if there's no comment, it should be empty
          if (!commentResult && !testRegExp(rest, emptyRegExp)) {
            return expected(rest, 'End of input');
          }

          return makeSuccess<LineNode>(
            {
              kind: NodeKind.line,
              range: [r.value.range[0], rest.index],
              contents: r.value,
              comment: commentResult?.value,
            },
            rest,
          );
        })(result);
        if (lineResultWithComment.kind === ResultKind.Error) {
          errors.push(lineResultWithComment);
        } else {
          lineNodes.push(lineResultWithComment.value);
        }
      }
    }

    index += line.length + 1;
  }

  return {
    errors,
    program: {
      kind: NodeKind.program,
      lines: lineNodes,
      range:
        lineNodes.length >= 1
          ? [lineNodes[0]!.range[0], lineNodes[lineNodes.length - 1]!.range[1]]
          : [0, sourceNormalized.length],
    },
    originalLines: lines,
  };
}
