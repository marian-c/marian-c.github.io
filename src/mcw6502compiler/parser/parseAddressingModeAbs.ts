import { expected, type StringInput, trimStart } from '@/mcw6502compiler/parser/stringInput';
import { type FinalConfig } from '@/mcw6502compiler/config';
import {
  makeError,
  makeSuccess,
  type Result,
  ResultKind,
} from '@/mcw6502compiler/parser/parserTypes';
import {
  type AbsAddressingModeNode,
  type InstructionOperatorNode,
  NodeKind,
  ValueLiteralKind,
} from '@/mcw6502compiler/parser/ast';
import { parseValuable } from '@/mcw6502compiler/parser/parseValuable';

export function parseAddressingModeAbs(
  stringInput: StringInput,
  config: FinalConfig,
  instruction: InstructionOperatorNode,
): Result<AbsAddressingModeNode> {
  const suffixValue = instruction.forcedTo;
  if (suffixValue === 'b') {
    return makeError(
      'ABS addressing mode can not be forced to byte',
      instruction.range,
      stringInput.meta.originalLines,
    );
  }
  const trimmed = trimStart(stringInput);

  const valuableResult = parseValuable(trimmed, config);
  if (valuableResult.kind === ResultKind.Error) {
    return valuableResult;
  }

  // abs syntax is a subset of abx or aby
  const rest = valuableResult.rest;
  if (rest.value.startsWith(',')) {
    return expected(rest, 'end of ABS mode');
  }

  const valuable = valuableResult.value;
  let hexLength = null;
  if (
    valuable.kind === NodeKind.valueLiteral &&
    valuable.valueLiteralKind === ValueLiteralKind.hex
  ) {
    hexLength = valuable.originalStringValue.length;
  }
  if (hexLength === 2 || hexLength === 1) {
    return makeError(
      'Hex literals must not have length 1 or 2 for ABS mode',
      [valuable.range[0] + 1, valuable.range[1]],
      stringInput.meta.originalLines,
    );
  }

  return makeSuccess<AbsAddressingModeNode>(
    {
      kind: NodeKind.absAddressingMode,
      value: valuableResult.value,
      range: [trimmed.index, rest.index],
    },
    valuableResult.rest,
  );
}
