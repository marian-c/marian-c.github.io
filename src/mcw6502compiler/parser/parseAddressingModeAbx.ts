import {
  expected,
  type StringInput,
  substr,
  trimStart,
} from '@/mcw6502compiler/parser/stringInput';
import { type FinalConfig } from '@/mcw6502compiler/config';
import {
  makeError,
  makeSuccess,
  type Result,
  ResultKind,
} from '@/mcw6502compiler/parser/parserTypes';
import {
  type AbxAddressingModeNode,
  type InstructionOperatorNode,
  NodeKind,
  ValueLiteralKind,
} from '@/mcw6502compiler/parser/ast';
import { parseValuable } from '@/mcw6502compiler/parser/parseValuable';

export function parseAddressingModeAbx(
  stringInput: StringInput,
  config: FinalConfig,
  instruction: InstructionOperatorNode,
): Result<AbxAddressingModeNode> {
  const suffixValue = instruction.forcedTo;
  if (suffixValue === 'b') {
    return makeError(
      'ABX addressing mode can not be forced to byte',
      instruction.range,
      stringInput.meta.originalLines,
    );
  }
  const trimmed = trimStart(stringInput);

  const valuableResult = parseValuable(trimmed, config);

  if (valuableResult.kind === ResultKind.Error) {
    return valuableResult;
  }

  const particle = /^,[xX]/.test(valuableResult.rest.value);
  if (!particle) {
    return expected(valuableResult.rest, 'the ",X" part of ABX mode');
  }

  const valuable = valuableResult.value;
  let hexLength = null;
  if (
    valuable.kind === NodeKind.valueLiteral &&
    valuable.valueLiteralKind === ValueLiteralKind.hex
  ) {
    hexLength = valuable.originalStringValue.length;
  }
  if (hexLength === 2 || hexLength === 1) {
    return makeError(
      'Hex literals must not have length 1 or 2 for ABX mode',
      [valuable.range[0] + 1, valuable.range[1]],
      stringInput.meta.originalLines,
    );
  }

  const rest = substr(valuableResult.rest, 2);

  return makeSuccess<AbxAddressingModeNode>(
    {
      kind: NodeKind.abxAddressingMode,
      value: valuableResult.value,
      range: [trimmed.index, rest.index],
    },
    rest,
  );
}
