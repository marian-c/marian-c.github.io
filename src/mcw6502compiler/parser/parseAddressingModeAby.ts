import {
  expected,
  type StringInput,
  substr,
  trimStart,
} from '@/mcw6502compiler/parser/stringInput';
import { type FinalConfig } from '@/mcw6502compiler/config';
import {
  makeError,
  makeSuccess,
  type Result,
  ResultKind,
} from '@/mcw6502compiler/parser/parserTypes';
import {
  type AbyAddressingModeNode,
  type InstructionOperatorNode,
  NodeKind,
  ValueLiteralKind,
} from '@/mcw6502compiler/parser/ast';
import { parseValuable } from '@/mcw6502compiler/parser/parseValuable';

export function parseAddressingModeAby(
  stringInput: StringInput,
  config: FinalConfig,
  instruction: InstructionOperatorNode,
): Result<AbyAddressingModeNode> {
  const suffixValue = instruction.forcedTo;
  if (suffixValue === 'b') {
    return makeError(
      'ABY addressing mode can not be forced to byte',
      instruction.range,
      stringInput.meta.originalLines,
    );
  }
  const trimmed = trimStart(stringInput);

  const valuableResult = parseValuable(trimmed, config);

  if (valuableResult.kind === ResultKind.Error) {
    return valuableResult;
  }
  const particle = /^,[yY]/.test(valuableResult.rest.value);
  if (!particle) {
    return expected(valuableResult.rest, 'the ",Y" part of ABY mode');
  }

  const valuable = valuableResult.value;
  let hexLength = null;
  if (
    valuable.kind === NodeKind.valueLiteral &&
    valuable.valueLiteralKind === ValueLiteralKind.hex
  ) {
    hexLength = valuable.originalStringValue.length;
  }
  if (hexLength === 2 || hexLength === 1) {
    return makeError(
      'Hex literals must not have length 1 or 2 for ABY mode',
      [valuable.range[0] + 1, valuable.range[1]],
      stringInput.meta.originalLines,
    );
  }

  const rest = substr(valuableResult.rest, 2);

  return makeSuccess<AbyAddressingModeNode>(
    {
      kind: NodeKind.abyAddressingMode,
      value: valuableResult.value,
      range: [trimmed.index, rest.index],
    },
    rest,
  );
}
