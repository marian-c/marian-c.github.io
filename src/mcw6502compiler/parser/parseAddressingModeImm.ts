import { type StringInput, substr, trimStart } from '@/mcw6502compiler/parser/stringInput';
import { type FinalConfig } from '@/mcw6502compiler/config';
import {
  makeError,
  makeSuccess,
  type Result,
  ResultKind,
} from '@/mcw6502compiler/parser/parserTypes';
import { type ImmAddressingModeNode, NodeKind } from '@/mcw6502compiler/parser/ast';
import { parseValuable } from '@/mcw6502compiler/parser/parseValuable';

export function parseAddressingModeImm(
  stringInput: StringInput,
  config: FinalConfig,
): Result<ImmAddressingModeNode> {
  const trimmed = trimStart(stringInput);
  if (!trimmed.value.startsWith('#')) {
    return makeError(
      'Expected immediate addressing mode',
      [trimmed.index, trimmed.index + trimmed.value.length],
      trimmed.meta.originalLines,
    );
  }

  const value = parseValuable(substr(trimmed, 1), config);
  if (value.kind === ResultKind.Error) {
    return value;
  }

  return makeSuccess<ImmAddressingModeNode>(
    {
      kind: NodeKind.immAddressingMode,
      value: value.value,
      range: [trimmed.index, value.rest.index],
    },
    value.rest,
  );
}
