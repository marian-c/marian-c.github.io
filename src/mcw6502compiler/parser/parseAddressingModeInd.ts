import {
  expected,
  type StringInput,
  substr,
  trimStart,
} from '@/mcw6502compiler/parser/stringInput';
import { type FinalConfig } from '@/mcw6502compiler/config';
import {
  makeError,
  makeSuccess,
  type Result,
  ResultKind,
} from '@/mcw6502compiler/parser/parserTypes';
import { type IndAddressingModeNode, NodeKind } from '@/mcw6502compiler/parser/ast';
import { parseValuable } from '@/mcw6502compiler/parser/parseValuable';

export function parseAddressingModeInd(
  stringInput: StringInput,
  config: FinalConfig,
): Result<IndAddressingModeNode> {
  const trimmed = trimStart(stringInput);

  if (!trimmed.value.startsWith('(')) {
    return expected(trimmed, 'Indirect addressing mode');
  }

  const value = parseValuable(substr(trimmed, 1), config);

  if (value.kind === ResultKind.Error) {
    return value;
  }

  if (value.rest.value.startsWith(' ') || value.rest.value.startsWith('\t')) {
    return makeError(
      'Unexpected empty space, expected the ")" part of IND mode',
      [value.rest.index, value.rest.index + 1],
      stringInput.meta.originalLines,
    );
  }
  if (!value.rest.value.startsWith(')')) {
    return expected(value.rest, 'the ")" part of IND mode');
  }

  const rest = substr(value.rest, 1);

  return makeSuccess<IndAddressingModeNode>(
    {
      kind: NodeKind.indAddressingMode,
      value: value.value,
      range: [trimmed.index, rest.index],
    },
    rest,
  );
}
