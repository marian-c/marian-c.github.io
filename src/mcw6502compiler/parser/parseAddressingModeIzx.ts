import {
  expected,
  type StringInput,
  substr,
  trimStart,
} from '@/mcw6502compiler/parser/stringInput';
import { type FinalConfig } from '@/mcw6502compiler/config';
import {
  makeError,
  makeSuccess,
  type Result,
  ResultKind,
} from '@/mcw6502compiler/parser/parserTypes';
import { type IzxAddressingModeNode, NodeKind } from '@/mcw6502compiler/parser/ast';
import { parseValuable } from '@/mcw6502compiler/parser/parseValuable';

export function parseAddressingModeIzx(
  stringInput: StringInput,
  config: FinalConfig,
): Result<IzxAddressingModeNode> {
  const trimmed = trimStart(stringInput);

  if (!trimmed.value.startsWith('(')) {
    return expected(trimmed, 'x-indexed indirect mode');
  }

  const value = parseValuable(substr(trimmed, 1), config);

  if (value.kind === ResultKind.Error) {
    return value;
  }

  if (value.rest.value.startsWith(' ') || value.rest.value.startsWith('\t')) {
    return makeError(
      'Unexpected empty space, expected the ",X)" part of IZX mode',
      [value.rest.index, value.rest.index + 1],
      stringInput.meta.originalLines,
    );
  }

  if (!value.rest.value.startsWith(',x)') && !value.rest.value.startsWith(',X)')) {
    return expected(value.rest, 'the ",X)" part of IZX mode');
  }

  const rest = substr(value.rest, 3);

  return makeSuccess<IzxAddressingModeNode>(
    {
      kind: NodeKind.izxAddressingMode,
      value: value.value,
      range: [trimmed.index, rest.index],
    },
    rest,
  );
}
