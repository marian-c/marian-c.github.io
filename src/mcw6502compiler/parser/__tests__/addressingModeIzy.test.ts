import { parseSource } from '@/mcw6502compiler/parser/parse';
import { phase2 } from '@/mcw6502compiler/instructionMatrixToInfo';
const instructionsInfo = phase2({
  BRK: { IMP: 1 },
  ROR: { IMP: 2 },
  INX: { IMP: 3 },
  LDA: { aby: 6, abs: 4, abx: 5, imm: 6, ind: 7, izx: 8, izy: 9 },
});

function handle(s: string) {
  return parseSource(s, { instructionsInfo });
}

describe('IZY addressing mode', () => {
  test('simple ', () => {
    expect(handle('LDA ($a1),y ;')).toMatchSnapshot();
  });
  test('simple 2', () => {
    expect(handle('LdA ($01),Y;')).toMatchSnapshot();
  });

  test('with symbol', () => {
    const result = handle('LdA (SR),Y');
    expect(result).toMatchSnapshot();
  });
  test('with decimal', () => {
    const result = handle('LdA (12),Y');
    expect(result).toMatchSnapshot();
  });
  test('with binary', () => {
    const result = handle('LdA (%0101),Y');
    expect(result).toMatchSnapshot();
  });
  test('with ascii', () => {
    const result = handle('LdA ("a"),Y');
    expect(result).toMatchSnapshot();
  });
  test('with ascii - multiple', () => {
    const result = handle('LdA ("abc"),Y');
    expect(result).toMatchSnapshot();
  });
});

describe('IZY mode errors', () => {
  test('errors', () => {
    expect(
      handle(`
LDA ($a1 ),y
LDA ($a1), y
LDA ($a1) ,y
LDA ($a1),y,
LDA ($a1),h
            
          
LDA (#$a1),y
    `),
    ).toMatchSnapshot();
  });
});
