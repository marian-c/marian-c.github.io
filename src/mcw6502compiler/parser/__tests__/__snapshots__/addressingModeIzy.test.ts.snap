// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IZY addressing mode simple  1`] = `
{
  "errors": [],
  "originalLines": [
    "LDA ($a1),y ;",
  ],
  "program": {
    "kind": "p",
    "lines": [
      {
        "comment": {
          "kind": "c",
          "range": [
            12,
            13,
          ],
          "value": ";",
        },
        "contents": {
          "instructionKind": 1,
          "kind": "i",
          "operand": {
            "kind": "izy",
            "range": [
              4,
              11,
            ],
            "value": {
              "kind": "vl",
              "originalStringValue": "a1",
              "range": [
                5,
                8,
              ],
              "valueLiteralKind": "hex",
            },
          },
          "operator": {
            "kind": "or",
            "range": [
              0,
              3,
            ],
            "value": "lda",
          },
          "range": [
            0,
            11,
          ],
        },
        "kind": "l",
        "range": [
          0,
          13,
        ],
      },
    ],
    "range": [
      0,
      13,
    ],
  },
}
`;

exports[`IZY addressing mode simple 2 1`] = `
{
  "errors": [],
  "originalLines": [
    "LdA ($01),Y;",
  ],
  "program": {
    "kind": "p",
    "lines": [
      {
        "comment": {
          "kind": "c",
          "range": [
            11,
            12,
          ],
          "value": ";",
        },
        "contents": {
          "instructionKind": 1,
          "kind": "i",
          "operand": {
            "kind": "izy",
            "range": [
              4,
              11,
            ],
            "value": {
              "kind": "vl",
              "originalStringValue": "01",
              "range": [
                5,
                8,
              ],
              "valueLiteralKind": "hex",
            },
          },
          "operator": {
            "kind": "or",
            "range": [
              0,
              3,
            ],
            "value": "lda",
          },
          "range": [
            0,
            11,
          ],
        },
        "kind": "l",
        "range": [
          0,
          12,
        ],
      },
    ],
    "range": [
      0,
      12,
    ],
  },
}
`;

exports[`IZY addressing mode with ascii - multiple 1`] = `
{
  "errors": [],
  "originalLines": [
    "LdA ("abc"),Y",
  ],
  "program": {
    "kind": "p",
    "lines": [
      {
        "comment": undefined,
        "contents": {
          "instructionKind": 1,
          "kind": "i",
          "operand": {
            "kind": "izy",
            "range": [
              4,
              13,
            ],
            "value": {
              "kind": "vl",
              "originalStringValue": "abc",
              "range": [
                5,
                10,
              ],
              "valueLiteralKind": "ascii",
            },
          },
          "operator": {
            "kind": "or",
            "range": [
              0,
              3,
            ],
            "value": "lda",
          },
          "range": [
            0,
            13,
          ],
        },
        "kind": "l",
        "range": [
          0,
          13,
        ],
      },
    ],
    "range": [
      0,
      13,
    ],
  },
}
`;

exports[`IZY addressing mode with ascii 1`] = `
{
  "errors": [],
  "originalLines": [
    "LdA ("a"),Y",
  ],
  "program": {
    "kind": "p",
    "lines": [
      {
        "comment": undefined,
        "contents": {
          "instructionKind": 1,
          "kind": "i",
          "operand": {
            "kind": "izy",
            "range": [
              4,
              11,
            ],
            "value": {
              "kind": "vl",
              "originalStringValue": "a",
              "range": [
                5,
                8,
              ],
              "valueLiteralKind": "ascii",
            },
          },
          "operator": {
            "kind": "or",
            "range": [
              0,
              3,
            ],
            "value": "lda",
          },
          "range": [
            0,
            11,
          ],
        },
        "kind": "l",
        "range": [
          0,
          11,
        ],
      },
    ],
    "range": [
      0,
      11,
    ],
  },
}
`;

exports[`IZY addressing mode with binary 1`] = `
{
  "errors": [],
  "originalLines": [
    "LdA (%0101),Y",
  ],
  "program": {
    "kind": "p",
    "lines": [
      {
        "comment": undefined,
        "contents": {
          "instructionKind": 1,
          "kind": "i",
          "operand": {
            "kind": "izy",
            "range": [
              4,
              13,
            ],
            "value": {
              "kind": "vl",
              "originalStringValue": "0101",
              "range": [
                5,
                10,
              ],
              "valueLiteralKind": "bin",
            },
          },
          "operator": {
            "kind": "or",
            "range": [
              0,
              3,
            ],
            "value": "lda",
          },
          "range": [
            0,
            13,
          ],
        },
        "kind": "l",
        "range": [
          0,
          13,
        ],
      },
    ],
    "range": [
      0,
      13,
    ],
  },
}
`;

exports[`IZY addressing mode with decimal 1`] = `
{
  "errors": [],
  "originalLines": [
    "LdA (12),Y",
  ],
  "program": {
    "kind": "p",
    "lines": [
      {
        "comment": undefined,
        "contents": {
          "instructionKind": 1,
          "kind": "i",
          "operand": {
            "kind": "izy",
            "range": [
              4,
              10,
            ],
            "value": {
              "kind": "vl",
              "originalStringValue": "12",
              "range": [
                5,
                7,
              ],
              "valueLiteralKind": "dec",
            },
          },
          "operator": {
            "kind": "or",
            "range": [
              0,
              3,
            ],
            "value": "lda",
          },
          "range": [
            0,
            10,
          ],
        },
        "kind": "l",
        "range": [
          0,
          10,
        ],
      },
    ],
    "range": [
      0,
      10,
    ],
  },
}
`;

exports[`IZY addressing mode with symbol 1`] = `
{
  "errors": [],
  "originalLines": [
    "LdA (SR),Y",
  ],
  "program": {
    "kind": "p",
    "lines": [
      {
        "comment": undefined,
        "contents": {
          "instructionKind": 1,
          "kind": "i",
          "operand": {
            "kind": "izy",
            "range": [
              4,
              10,
            ],
            "value": {
              "kind": "r",
              "range": [
                5,
                7,
              ],
              "referencedName": "sr",
            },
          },
          "operator": {
            "kind": "or",
            "range": [
              0,
              3,
            ],
            "value": "lda",
          },
          "range": [
            0,
            10,
          ],
        },
        "kind": "l",
        "range": [
          0,
          10,
        ],
      },
    ],
    "range": [
      0,
      10,
    ],
  },
}
`;

exports[`IZY mode errors errors 1`] = `
{
  "errors": [
    {
      "kind": 1,
      "lineRange": {
        "end": [
          2,
          10,
        ],
        "start": [
          2,
          9,
        ],
      },
      "message": "Unexpected empty space, expected the "),Y" part of IZY mode",
      "range": [
        9,
        10,
      ],
    },
    {
      "kind": 1,
      "lineRange": {
        "end": [
          3,
          11,
        ],
        "start": [
          3,
          10,
        ],
      },
      "message": "Unexpected identifier: ,, expected End of input",
      "range": [
        23,
        24,
      ],
    },
    {
      "kind": 1,
      "lineRange": {
        "end": [
          4,
          13,
        ],
        "start": [
          4,
          11,
        ],
      },
      "message": "Unexpected identifier: ,y, expected End of input",
      "range": [
        37,
        39,
      ],
    },
    {
      "kind": 1,
      "lineRange": {
        "end": [
          5,
          13,
        ],
        "start": [
          5,
          12,
        ],
      },
      "message": "Unexpected identifier: ,, expected End of input",
      "range": [
        51,
        52,
      ],
    },
    {
      "kind": 1,
      "lineRange": {
        "end": [
          6,
          12,
        ],
        "start": [
          6,
          10,
        ],
      },
      "message": "Unexpected identifier: ,h, expected End of input",
      "range": [
        62,
        64,
      ],
    },
    {
      "kind": 1,
      "lineRange": {
        "end": [
          9,
          13,
        ],
        "start": [
          9,
          6,
        ],
      },
      "message": "Unexpected identifier: #$a1),y, expected a valuable",
      "range": [
        94,
        101,
      ],
    },
  ],
  "originalLines": [
    "",
    "LDA ($a1 ),y",
    "LDA ($a1), y",
    "LDA ($a1) ,y",
    "LDA ($a1),y,",
    "LDA ($a1),h",
    "            ",
    "          ",
    "LDA (#$a1),y",
    "    ",
  ],
  "program": {
    "kind": "p",
    "lines": [],
    "range": [
      0,
      106,
    ],
  },
}
`;
